### ISSUE TITLE
Using an Intrinsic Function in !Ref with LanguageExtensions and Serverless Transform results in an error

### ISSUE THREAD

**Issue by georgealton** [aws-cloudformation/cfn-lint#4556]
### CloudFormation Lint Version

cfn-lint 1.52.0

### What operating system are you using?

fedora

### Describe the bug

I'd like to build a Ref to a LogicalId using Parameters to a Template. I'm using this with a SAM / Serverless Transform.

when running cfn-lint against a template, cfn-lint fails to transform / decode the template.

as far as I can tell my template adheres to the constraints in 

Supported Functions in Ref with LanguageExtensions
https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/intrinsic-function-reference-ref.html#ref-supported-functions

SAM and Language Extensions
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-language-extensions.html

Using !Ref & Fn::Sub with Serverless Resources instead of SAM notation
https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/transform-aws-serverless.html#aws-serverless-language-extensions

### Expected behavior

When using an Intrinsic Function with LanguageExtensions and Serverless Transforms the template should not fail to transform.

### Reproduction template

Here's the minimal reproducible example

`test.yaml`

```yaml
Transform:
  - "AWS::LanguageExtensions"
  - "AWS::Serverless-2016-10-31"
Parameters:
  Environment:
    Type: String
    Default: "dev"
  BucketNamedev:
    Type: String
    Default: "test"
Resources:
  Bucket:
    Type: AWS::S3::Bucket
    Properties:
      BucketName: !Ref
        Fn::Sub: "BucketName${Environment}"
```

```console
$ cfn-lint -t test.yaml
E0001 Error transforming template: cannot use 'dict' as a dict key (unhashable type: 'dict')
test.yaml:1:1
```

the same issue with `Ref:` instead of the Tag `!Ref`

```yaml
Transform:
  - "AWS::LanguageExtensions"
  - "AWS::Serverless-2016-10-31"
Parameters:
  Environment:
    Type: String
    Default: "dev"
  BucketNamedev:
    Type: String
    Default: "test"
Resources:
  Bucket:
    Type: AWS::S3::Bucket
    Properties:
      BucketName:
        Ref:
          Fn::Sub: "BucketName${Environment}"
```

```console
$ cfn-lint -t test.yaml
E0001 Error transforming template: cannot use 'cfnlint.decode.node.create_dict_node_class.<locals>.node_class' as a dict key (unhashable type: 'dict_node')
test.yaml:1:1
```

As an experiment, Without the Serverless Transform;
```yaml
Transform:
  - "AWS::LanguageExtensions"
  # - "AWS::Serverless-2016-10-31"
Parameters:
  Environment:
    Type: String
    Default: "dev"
  BucketNamedev:
    Type: String
    Default: "test"
Resources:
  Bucket:
    Type: AWS::S3::Bucket
    Properties:
      BucketName:
        Ref:
          Fn::Sub: "BucketName${Environment}"
```

removing the serverless transform does not error, In my real Templates I have `AWS::Serverless::*` Resource Types, so this wouldn't be an acceptable solution for me.

```console
$ cfn-lint -t test.yaml
$ echo $?
0
```
